README.md
# round-py: A Python client for the Gem API
The round client is designed to interact with Gem's API to make building blockchain apps drop dead simple.  All the complexity of altcoin protocols and cryptography has been abstracted away so you can focus on building your product.  Here are a few of the many great things the API and clients provide:

* Multi-signature wallets with Gem as a cosigner
* Bitcoin, Testnet, Litecoin, and Dogecoin support (multisig for all!)
* Webhook notifications automatically subscribed for you
* Integrated 2FA solution with arbitrary endpoints to build into your app
* Simplified balance inqueries
* Easy address management
* Hardware Security Modules for co-signing key
* Rules engine for transactions
* SDKs for many popular languages

## Support information
* __Support email__: [support@gem.co](mailto:support@gem.co)
* __Issues__:  Use github issues
* __Support Slack room__:  [![](https://chat.gem.co/badge.svg)](https://chat.gem.co)
* __Detailed API Docs__:  [Gem API Docs](http://guide.gem.co)

## Installing round-py:
### Prerequisites:
* Python 2.7/3.3+
* Git
* [virtualenv](http://pypi.python.org/pypi/virtualenv)/[virtualenvwrapper](https://virtualenvwrapper.readthedocs.org/en/latest/) or a similar virtual environment solution is recommended. (`sudo` is bad)

#### [Linux (debian-based, tested on Ubuntu 14.04)](docs/install.md#linux-debian-based-tested-on-ubuntu-1404)
#### [Mac OSX](docs/install.md#mac-osx)
#### [Windows](docs/install.md#windows)
#### [Heroku](docs/install.md#heroku)

## Getting Started Tutorial
#### Table of Contents
* [Introduction](README.md#Introduction)
* [1. Run the client](README.md#1-run-the-client)
* [2. Configure your application and API token](README.md#2-configure-your-application-and-api-token)
* [3. Create your User and Wallet](README.md#3-create-your-user-and-wallet)
* [4. Authenticate your User](README.md#4-authenticate-your-user)
* [5. Access the wallet and Default Account](README.md#5-access-the-wallet-and-default-account)
* [6. Generate an Address and Add Funds](README.md#6-generate-an-address-and-add-funds)
* [7. Make a Payment](README.md#7-make-a-payment)
* [Advanced Topics](docs/advanced.md)
    * [More about Wallets and Accounts](docs/advanced.md#wallets-and-accounts)
    * [More about Transactions](docs/advanced.md#transactions-and-payments)
    * [Subscriptions](docs/advanced.md#subscriptions)
    * [Integrated 2FA](docs/advanced.md#integrated-2fa)
    * [Operational/Custodail wallet models](docs/advanced.md#operationalcustodial-wallets)
    * [Operational/Custodial payments](docs/advanced.md#payments)

### Introduction
This tutorial will have you run through setting up your application and creating your own wallet as a user of your application.  By the end of the tutorial, you will have created your User, wallet, account, an address as well as fund it and then make a payment using the bitcoin testnet network.

This tutorial assumes that you have completed the developer signup and that you have successfully [installed the client](docs/install.md)

### 1. Run the Client
In this step you will learn how to instantiate the API client for the given networks.

1. start your favorite interactive shell and import the round library

    ```bash
    $ python
    >>> import round
    # If you need access to the round builtin, then import as something else:
    >>> import round as gemlib
    # Or just import the client factory:
    >>> from round import client
    ```

1. Create the client object using the sandbox environment

    ```python
    # the default client is set to sandbox (the testnet environment)
    client = round.client()
    ```

[[top]](README.md#getting-started-tutorial)

### 2. Configure your applicaiton and API Token
In this step your application and you will retrieve the API Token for the application and set your applications redirect uri.  The uri is used to push the user back to your app after they complete an out of band challenge.

1. Set the redirect uri by clicking in the options gear and selecting `add redirect_uri`

1. In the [console](https://sandbox.gem.co) click `show` to copy your api_token.

1. Go back to your shell session and set a variable for api_token.

    ```python
    api_token = u'q234t09ergoasgr-9_qt4098qjergjia-asdf2490'
    ```

[[top]](README.md#getting-started-tutorial)

### 3. Create your User and Wallet
In this step you will create your own personal Gem user and wallet authorized on your application.  This is an end-user account, which will have a 2-of-3 multisig bitcoin wallet.

1. Create your user and wallet:

    ```python
    #  Store the device token for future authentication
    device_token = client.users.create(
        first_name = "YOUR FIRST NAME",
        last_name = "YOUR LAST NAME",
        email = "YOUR EMAIL ADDRESS",
        passphrase = "aReallyStrongPassphrase",
        device_name = "SOME DEVICE NAME",
        api_token = "YOUR API TOKEN",
        redirect_uri = "http://optional-uri/user-device-approved")
    ```

1. Your application should **store the device_token permanently** as this will be required to authenticate from your app as this user. If you lose it, you will have to create a new device for the user to authorize. (`device_token, mfa_uri = user.devices.create(name)` - the mfa_uri is where you redirect the user to provide their confirmation token.)
1. You will receive an email from Gem asking you to confirm your account and finish setup.  Please follow the instructions. At the end of the User sign up flow, you'll be redirected to the redirect_uri provided in users.create (if you provided one).

[[top]](README.md#getting-started-tutorial)

### 4. Authenticate your User
In this step you will learn how to authenticate to the Gem API on a User's device to get a fully functional User object with which to perform wallet actions.

1. Call the authenticate_device method from the client object

    ```python
    full_user = client.authenticate_device(
        api_token = api_token,
        device_token = device_token,
        email = email)
    ```

[[top]](README.md#getting-started-tutorial)

### 5. Access the wallet and Default Account
[Wallets and Accounts](docs/advanced.md#wallets-and-accounts)

1. Get the default wallet and then default account (which is a bitcoin account)

    ```python
    my_wallet = full_user.wallet
    my_account = my_wallet.accounts['default']
    ```

2. Or create an account...say an altcoin!

    ```python
    #network takes one of 'bitcoin', 'bitcoin_testnet', 'litecoin', 'dogecoin'
    my_account = my_wallet.accounts.create(name='brand-new-account', network='litecoin')
    ```

[[top]](README.md#getting-started-tutorial)

### 6. Generate an Address and Add Funds
In this section you'll learn how to create an address to fund with testnet coins aka funny money.

1. Create an address

    ```python
    address = my_account.addresses.create()
    print address.string, address.path
    ```
1. Copy the address string and go to a faucet to fund it:
    1. [TP's Testnet Faucet](https://tpfaucet.appspot.com/)
    1. [Haskoin Testnet Faucet](http://faucet.haskoin.com/)

Payments have to be confirmed by the network and on Testnet that can be slow.  To monitor for confirmations: input the address into the following url `https://live.blockcypher.com/btc-testnet/address/<YOUR ADDRESS>`.  The current standard number of confirmations for a transaction to be considered safe is 6.

You will be able to make a payment with only one confirmation, however.  While you wait for that to happen, feel free to read more details about:
[Wallets and Accounts](docs/advanced.md#wallets-and-accounts)

[[top]](README.md#getting-started-tutorial)

### 7. Make a Payment
In this section you’ll learn how to create a payment a multi-signature payment in an HD wallet.  Once your address gets one confirmation we’ll be able to send a payment out of the wallet.  To make a payment, you'll unlock a wallet, generate a list of payees and then call the pay method.

1. Unlock the wallet:

    ```python
    my_wallet.unlock(<PASSPHRASE>) # Same passphrase used in users.create
    ```
1. Make a payment

    ```python
    transaction = my_account.pay(
        payees=[{'address': 'mxzdT4ShBudVtZbMqPMh9NVM3CS56Fp11s', 'amount':25000}],
        utxo_confirmations = 1,
        redirect_uri='https://my.mobileapp.com/user_redirect')
    ```
1. Direct the user to the Gem multi-factor authentication(MFA) url for user to confirm the tx

   ```python
   transaction.attributes[u'mfa_url']
   ```

The pay call takes a list of payee objects.  A payee is a dict of `{'address':ADDRESS, 'amount':amount}` where address is the bitcoin address and amount is the number of satoshis.  `utxo_confirmations` default to 6 and represents the number of confirmations an unspent output needs to have in order to be selected for the transaction.  The last argument is the redirect uri for Gem to send the user back to your application after the user submits their MFA challenge.

**CONGRATS** - now build something cool.

[[top]](README.md#getting-started-tutorial)


CHANGELOG.md

0.9.8 / 2015-09-02
==================

  * Support change_account in wallet.pay and account.pay
  * Minor improvements: normalized ListWrapper accessor behavior, etc...
  * Optimizations: use account_query and wallet_query on collection access rather than making a list call

0.9.7 / 2015-08-25
==================

  * Do not populate collections by default - populate them on repr, iter, refresh, or if populate is supplied on creation

0.9.5 / 2015-08-20
==================

  * Added BalanceError
  * Minor bugfixes

0.9.4 / 2015-08-17
==================

  * Support for application wallet management through the CLI
  * Wallet.pay now supports paying from all accounts without specifying explicit payer values
  * Support application wallets in the CLI
  * Added `balances_at` which queries for balance as a function of required confirmations
  * Added support for wallet and account_query (to reduce unnecessary list calls)
  * Added netki support!
  * Backporting cacheable from 0.10.0
  * Replaced print statements with logging
  * Minor bugfixes

0.9.3 / 2015-08-05
==================

  * provider user, wallet, and account calls that accept a key parameter for notification convenience
  * pad pyotp tokens to six characters

0.9.2 / 2015-07-24
==================

  * Default to encrypting with AES+SHA256HMAC - NaCl will be removed in 0.9.3
  * Make with_mfa accept callables from anywhere
  * Memoize addresses - added account.get_addresses to provide an optional refresh (fetch) parameter

0.9.1 / 2015-07-14
==================

  * wallet.pay exists now

0.9.0 / 2015-06-04
==================

  * Python3.3+ compatible!
  * Support Gem's new key encryption scheme (no more libsodium!)

0.8.3 / 2015-06-25
==================

  * Accept an optional encrypted_seed parameter in unlock to allow applications to store their own primary private keys.

0.8.2 / 2015-06-19
==================

  * Bugfix: Use NaclPassphraseBox to create users for 0.8.x (was using the ppbox from 0.9.x)

0.8.1 / 2015-06-17
==================

  * Backporting py3 support from 0.9.0
  * Allow `account.pay` to take a callable instead of a string for `mfa_token`
  * Support old and new encryption schemes

0.8.0 / 2015-05-26
==================

  * support for litecoin and dogecoin!
  * removed network references and disabled sig_hash checking until we can eliminate python-bitcoinlib in coinop
  * removed old user.wallets accessor
  * various improvements and bug fixes
  * examples/cli.py opens a browser for you when necessary

0.7.4 / 2015-05-13
==================

  * various improvements and bug fixes
  * new Error classes in the public interface

0.7.3 / 2015-04-29
==================

  * Added simple little CLI sample project
  * Updated client to work with the new authorization_request flow
  * Reordered some parameters for consistency
  * Users are limited to one wallet. (accessible through user.wallet)
  * account.transactions now accepts a list of acceptable statuses as filters
  * Various minor improvements and bug fixes


0.7.2 / 2015-04-17
==================

  * Proper handling of Application wallets
  * Added user.devices
  * Added send_mfa and verify_mfa to User objects, for arbitrary MFA through Gem
  * instance_id/instance_token normalized to new name: admin_token
  * Cleaned up the context parameter storage in Context
  * Various minor improvements and bug fixes

0.7.1 / 2015-04-02
==================

  * Updated with the live sandbox URL!
  * Minor tweaks

0.7.0 / 2015-04-01
==================

  * TOTP support for Applications
  * Wallets.generate can provide arbitrary trees for a wallet
  * A _useful_ README
  * Docstrings!
  * New transaction flow supporting MFA
  * Eliminated the 'payments' abstraction
  * Backup seed generation for User accounts removed
  * Various bug fixes and minor improvements

0.6.2 / 2015-03-04
==================

  *  PyNaCl 0.3.0 is released :)
  *  Update README with easier install instructions
  *  Begin documentation for public release
  *  Minor bugfixes

0.6.1 / 2015-03-02
==================

  *  tweak to linux install instructions
  *  Added subscription support (event notifications)
  *  Added confirmations parameter to account.pay
  *  updated readme with better linux installation instructions

0.6.0 / 2015-02-16
==================

  * Renamed users.create(wallet to users.create(default_wallet
  * Cleanup
  * Added client-side unit tests

0.5.5 / 2015-01-27
==================

  * Moved begin_ and complete_device_authorization to the client object and require the user email address as a parameter (also renamed `name` argument to `device_name` for clarity)
  * Updated README with Heroku/Mac OS X install instructions

0.5.4 / 2015-01-12
==================

  * Remove privkey from parameters to prevent patchboard from sending it in dev.create requests
  * Fix for setting mainnet params in pycoin

0.5.3 / 2014-12-29
==================

  * Set default sandbox url to api-sandbox.gem.co
  * Added exception for locked wallets
  * added client.developer test
  * added test_authenticate_otp
  * Added authenticate_device
  * removed unnecessary default credential from otp
  * added authenticate_application -- packaged client tests will minimally interact with the API and will be limited to unit tests. Functional client testing will be maintained in-house (as it requires access to a development instance of the API)

round/wallets.py
# -*- coding: utf-8 -*-
# wallets.py
#
# Copyright 2014 BitVault, Inc. dba Gem

from __future__ import unicode_literals
from future.utils import iteritems
import logging

from .config import *

from coinop.passphrasebox import PassphraseBox
from coinop.crypto.passphrasebox import PassphraseBox as NaclPassphraseBox

from coinop.multiwallet import MultiWallet
from coinop.transaction import Transaction as CoinopTx

from .wrappers import *
from .errors import *
from .accounts import Account, Accounts
from .subscriptions import Subscription, Subscriptions
import round.transactions as txs

logger = logging.getLogger(__name__)

def generate(passphrase, trees=['primary']):
    """Generate a seed for the primary tree of a Gem wallet.

    You may choose to store the passphrase for a user so the user doesn't have
    to type it in every time. This is okay (although the security risks should
    be obvious) but Gem strongly discourages storing even the encrypted private
    seed, and storing both the passphrase and the private seed is completely
    insane. Don't do it.

    Args:
      passphrase (str): The passphrase that will be used to encrypt the seed
        before it's send to Gem. Key-stretching is done with PBDKF2 and
        encryption is done with nacl's SecretBox.
      trees (list of str): A list of names to generate trees for. For User
        Wallets this will be ['primary'], for Application Wallets it will be
       ['primary', 'backup'].

    Returns:
      A dict of dicts containing the serialized public master node, and
       a sub-dict with the encrypted private seed for each tree in `trees`.
    """
    seeds, multi_wallet = MultiWallet.generate(trees, entropy=True)

    result = {}
    for tree in trees:
        result[tree] = dict(private_seed=seeds[tree],
                            public_seed=multi_wallet.public_wif(tree),
                            encrypted_seed=PassphraseBox.encrypt(passphrase,
                                                                 seeds[tree]))
    return result


class Wallets(DictWrapper):
    """A collection of round.Wallets objects."""

    def __init__(self, resource, client, populate=False, application=False):
        # This is less than awesome. Ideally a PB resource can learn whether it's
        # an application_wallets or user_wallets object.
        self.application = application
        super(Wallets, self).__init__(resource, client, populate)

    def __getitem__(self, name):
        if name in self._data: return self._data.__getitem__(name)
        try:
            wallet = self.wrap(
                self.application.resource.wallet_query(dict(name=name)).get())
            self.add(wallet)
            return wallet
        except Exception as e:
            logger.debug(e)
            raise KeyError(name)

    def create(self, name, passphrase=None, wallet_data=None):
        """Create a new Wallet object and add it to this Wallets collection.
        This is only available in this library for Application wallets. Users
        must add additional wallets in their User Console

        Args:
          name (str): wallet name
          passphrase (str, optional): A passphrase with which to encrypt a user
            wallet. If not supplied, wallet_data is mandatory.
          wallet_data (dict): Output from wallets.generate.
            For User Wallets, only the primary tree is used.
            For Application Wallets, the primary and backup trees are used.

        Returns:
          A tuple of the (backup_private_seed, round.Wallet).
        """
        if not self.application:
            raise RoundError("User accounts are limited to one wallet. Make an "
                             "account or shoot us an email <dev@gem.co> if you "
                             "have a compelling use case for more.")
        if not passphrase and not wallet_data:
            raise ValueError("Usage: wallets.create(name, passphrase [, "
                             "wallet_data])")
        elif passphrase:
            wallet_data = generate(passphrase,
                                   trees=(['primary', 'backup'] if (
                                       self.application) else ['primary']))

        wallet = dict(
            primary_private_seed=wallet_data['primary']['encrypted_seed'],
            primary_public_seed=wallet_data['primary']['public_seed'],
            name=name)
        if self.application:
            wallet['backup_public_seed'] = wallet_data['backup']['public_seed']

        resource = self.resource.create(wallet)
        wallet = self.wrap(resource)
        return (wallet_data['backup']['private_seed'], self.add(wallet)) if (
            self.application) else self.add(wallet)

    def wrap(self, resource):
        return Wallet(resource, self.client, self.application)


class Wallet(Wrapper, Updatable):
    """A Gem Wallet represents a 3-key multisig HD bitcoin wallet.

    Attributes:
      primary_private_seed (dict): An encrypted representation of the private
        seed for the primary tree. This is encrypted with a user-chosen
        passphrase that is never sent to Gem. This is used for signing
        transactions within this client libary.
      primary_public_seed (str): A misnomer, this is the serialized master node
        of the primary tree.
      backup_public_seed (str): A misnomer, this is the serialized master node
        of the backup tree. The backup private seed is generated out-of-band by
        the end user when they confirm their email address and MFA information.
        In the case of an Application wallet, the backup public can be delivered
        in-band during the wallets.create call.
      cosigner_public_seed (str): A misnomer, this is the serialized master node
        of the cosigner tree. The cosigner private seed is generated and held by
        Gem in Hardware Security Modules.
      name (str): A human-readable name for the wallet. A User's primary wallet
        is named 'default'.
      multi_wallet (coinop.Transaction): When unlocked, this variable will be
        populated with a high-level interface wrapper around the primary private
        seed.

    Args:
      resource (patchboard.Resource): server-side Wallet object
      client (round.Client)
    """

    def __init__(self, resource, client, application=None):
        super(Wallet, self).__init__(resource, client)

        self.application = application
        self.multi_wallet = None

    def is_unlocked(self):
        """Return true if the wallet is unlocked."""
        return not self.is_locked()

    def is_locked(self):
        """Return true if the wallet is locked."""
        return (self.multi_wallet is None)

    def unlock(self, passphrase, encrypted_seed=None):
        """Unlock the Wallet by decrypting the primary_private_seed with the
        supplied passphrase. Once unlocked, the private seed is accessible in
        memory and calls to `account.pay` will succeed. This is a necessary step
        for creating transactions.

        Args:
          passphrase (str): The passphrase the User used to encrypt this wallet.
          encrypted_seed (dict): A dictionary of the form
            {'ciphertext': longhexvalue,
             'iterations': integer of pbkdf2 derivations,
             'nonce': 24-byte hex value
             'salt': 16-byte hex value}
            this dict represents an private seed (not a master key) encrypted
            with the `passphrase` using pbkdf2. You can obtain this value with
            wallet.generate. If this value is supplied, it overwrites (locally
            only) the encrypted primary_private_seed value, allowing you to load
            in a primary key that you didn't store with Gem. Note that the key
            MUST match the pubkey that this wallet was created with.
        Returns:
          self
        """
        wallet = self.resource
        if not encrypted_seed:
            encrypted_seed = wallet.primary_private_seed
        try:
            if encrypted_seed['nonce']:
                primary_seed = NaclPassphraseBox.decrypt(
                    passphrase, encrypted_seed)
            else:
                primary_seed = PassphraseBox.decrypt(
                    passphrase, encrypted_seed)
        except:
            raise InvalidPassphraseError()

        self.multi_wallet = MultiWallet(
            private_seeds={'primary': primary_seed},
            public={'cosigner': wallet.cosigner_public_seed,
                    'backup': wallet.backup_public_seed})
        return self

    def balances_at(self, utxo_confirmations=6, network=None):
        """Return the confirmed, claimed (reserved for a pending, unsigned
        transaction), and available balances, where the threshold for
        confirmed is the value of `utxo_confirmations`.

        Args:
          utxo_confirmations (int): the # of confirmations to use when computing
            balances.
          network (str): Type of cryptocurrency.  Can be one of, 'bitcoin', '
            bitcoin_testnet', 'litecoin', 'dogecoin'.

        Returns:
          A dict of form { u'available_balance': 0,
                           u'claimed_balance': 0,
                           u'confirmed_balance': 0,
                           u'utxo_confirmations': 0 }
        """
        if not network in SUPPORTED_NETWORKS:
            raise ValueError('Network not valid!')
        return self.resource.available({'utxo_confirmations': utxo_confirmations,
                                        'network': network}).__dict__['data']

    @property
    @cacheable
    def accounts(self):
        """Return the cached Accounts object for this Wallet."""
        return self.get_accounts()

    def get_accounts(self, fetch=False):
        """Return this Wallet's accounts object, populating it if fetch is True."""
        return Accounts(self.resource.accounts, self.client, wallet=self, populate=fetch)

    @property
    def default_account(self):
        return self.accounts['default']

    def account(self, key=None, address=None, name=None):
        """Query for an account by key, address, or name."""
        if key:
            return self.client.account(key, wallet=self)
        if address:
            q = dict(address=address)
        elif name:
            q = dict(name=name)
        else:
            raise TypeError("Missing param: key, address, or name is required.")

        return Account(
            self.resource.account_query(q).get(), self.client, wallet=self)

    def dump_addresses(self, network, filename=None):
        """Return a list of address dictionaries for each address in all of the
        accounts in this wallet of the network specified by `network`
        """
        addrs =  [addr.data for a in self.accounts.values() if a.network == network
                            for addr in a.addresses]
        if filename:
            from json import dump
            with open(filename, 'w') as f:
                dump(addrs, f)

        return addrs

    @property
    @cacheable
    def subscriptions(self):
        """Return the cached Subscriptions object for this Wallet."""
        return self.get_subscriptions()

    def get_subscriptions(self, fetch=False):
        """Return this Wallet's subscriptions object, populating it if fetch is True."""
        return Subscriptions(
            self.resource.subscriptions, self.client, populate=fetch)


    def _get_account_attr(self, obj, attr='key'):
        try:
            return obj.resource.attributes[attr]
        except AttributeError:
            return self.accounts[obj].resource.attributes[attr]

    def pay(self, payees, remainder_account, payers=None, change_account=None,
            network=None, utxo_confirmations=6, mfa_token=None, redirect_uri=None):
        """Create, verify, and sign a new Transaction.

        This method is distinct from Account.pay in that it exposes the `payers`
        parameter which allows you to create transactions using inputs from
        multiple Accounts (which must ALL belong to this Wallet) by specifying
        a dict with account names or Account objects and amounts to be deducted.

        Note that a change output will be created for each account if necessary
        to deduct the precise amount specified. Fees will be drawn from the
        `remainder_account`, as well as any difference between the sum of `payee`
        outputs and the `payers` inputs. (Which is why `remainder_account` is
        mandatory and `payers` is optional.)

        If this Wallet is owned by a User object, the user must be redirected to
        a URL (`mfa_uri`) returned by this call to input their MFA token. After
        they complete that step, the Transaction will be approved and published
        to the bitcoin network. If a `redirect_uri` is provided in this call, the
        user will be redirected to that uri after they complete the MFA challenge
        so it's a good idea to have an endpoint in your app (or custom scheme on
        mobile) that can provide the user a seamless flow for returning to your
        application. If they have not configured a TOTP MFA application (e.g.
        Google Authenticator), then an SMS will be sent to their phone number
        with their token.

        If this Wallet is owned by an Application, the `mfa_token` can be
        included in this call and the Transaction will be automatically approved
        and published to the blockchain.

        ********************************* NOTE *********************************

        Additionally, if your use-case does not require keeping precise values
        in your Accounts, you may want to create a transaction without limiting
        your UTXO selection to specific Accounts.

        To do this, you must explicitly pass in `None` as the value for
        `remainder_account` and omit the `payers` parameter. This is to prevent
        accidental spends out of incorrect accounts.

        ************************************************************************

        Args:
          payees (list of dict): list of outputs in the form:
            [{'amount': 10000(satoshis),
              'address':'validbtcaddress'}, ...]
          remainder_account (str or Account or None): an Account to handle the
            difference between payer and payee sums as well as tx fees
            (if set to None, a transaction will be created that potentially
             uses UTXOs from every Account belonging to this wallet)
          payers (list of dict, optional): list of input accounts in the form:
            [{'amount': 10000(satoshis),
              'account': ('accountname'||accountInstance)}, ...]
          change_account (str or Account): if supplied, this account will
            be used to generate a change address in the event that a change
            output is required. Note that this does not replace the change
            outputs that will be generated to ensure that each Account in
            `payers` (if supplied) is deducted the precise amounts specified.
          network (str): Type of cryptocurrency.  Can be one of, 'bitcoin', '
            bitcoin_testnet', 'litecoin', 'dogecoin'.
          utxo_confirmations (int, optional): Required confirmations for UTXO
            selection ( > 0)
          mfa_token (str/function, optional): TOTP token for the Application
            owning this Account's wallet OR a callable/function which will
            generate such a token. The latter is suggested
            (e.g. application.get_mfa) as otherwise, the token might be
            invalidated by the time tx.create and tx.update complete (before
            the tx.approve call which actually requires the mfa_token).
          redirect_uri (str, optional): URI to redirect a user to after they
            input an mfa token on the page referenced by the `mfa_uri` returned
            by this function.

        Returns: An "unapproved" Transaction with an `mfa_uri` attribute to route
          the user to the MFA confirmation page --  if called with Gem-Device
          authentication.
          An "unconfirmed" Transaction -- if called with Gem-Application auth
          (and an `mfa_token` was supplied).
        """
        # Check that wallet is unlocked
        if self.is_locked():
            raise DecryptionError("This wallet must be unlocked with "
                                  "wallet.unlock(passphrase)")

        # First create the unsigned tx.
        content = dict(payees=payees,
                       utxo_confirmations=utxo_confirmations,
                       network=network)

        if remainder_account is not None:
            if payers: raise TypeError(
                    "Invalid payers: either supply a remainder_account "
                    "or omit the payers parameter")
            content['network'] = self._get_account_attr(
                remainder_account, 'network')
            content['remainer_account'] = self._get_account_attr(
                remainder_account, 'key')

        elif not payers and not network:
            raise TypeError("Missing network: network is required if "
                            "remainder_account is not specified")

        if change_account:
            content['change_account'] = self._get_account_attr(change_account)

        if payers: content['payers'] = [
            dict(p, account=self._get_account_attr(p['account'])) for p in payers
        ]

        try:
            unsigned = self.resource.transactions().create(content)
        except ResponseError as e:
            if "cannot cover" in e.message:
                raise BalanceError(e.message)
            raise e

        # Sign the tx with the primary private key.
        coinoptx = CoinopTx(data=unsigned.attributes)
        signatures = self.signatures(coinoptx)

        # Update the tx with the signatures.
        transaction = dict(signatures=dict(inputs=signatures,
                                           transaction_hash=coinoptx.hash))
        if redirect_uri:
            transaction['redirect_uri'] = redirect_uri

        signed = txs.Transaction(unsigned.update(transaction), self.client)

        # If this is an Application wallet, approve the transaction.
        if mfa_token and self.application:
            if hasattr(mfa_token, '__call__'): # callable() is unsupported by 3.1 and 3.2
                mfa_token = mfa_token.__call__()
            try:
                return txs.Transaction(signed.with_mfa(mfa_token).approve(),
                                       self.client)
            except Exception as e:
                signed.cancel()
                logger.debug(e.message)
                logger.debug("If you are having trouble with MFA tokens, make "
                             "sure your system time is accurate with `date -u`!")

        # Otherwise return the unapproved tx (now redirect the user to the
        # `mfa_uri` attribute to approve!)
        return signed

    def signatures(self, transaction):
        """Sign a transaction.

        Args:
          transaction (coinop.Transaction)

        Returns:
          A list of signature dicts of the form
            [ {'primary': 'base58signaturestring'},
              ... ]
        """
        # TODO: output.metadata['type']['change']
        if not self.multi_wallet:
            raise DecryptionError("This wallet must be unlocked with "
                                  "wallet.unlock(passphrase)")

        return self.multi_wallet.signatures(transaction)
